apps/webservice/src/app/[workspaceSlug]/_components/EditTarget.tsx

```diff
@@ -0,0 +1,289 @@
+"use client";
+
+import type { Target } from "@ctrlplane/db/schema";
+import React, { useState } from "react";
+import { useRouter } from "next/navigation";
+import { IconX } from "@tabler/icons-react";
+import yaml from "js-yaml";
+import { z } from "zod";
+
+import { Button } from "@ctrlplane/ui/button";
+import {
+  Dialog,
+  DialogContent,
+  DialogDescription,
+  DialogFooter,
+  DialogHeader,
+  DialogTitle,
+  DialogTrigger,
+} from "@ctrlplane/ui/dialog";
+import {
+  Form,
+  FormControl,
+  FormField,
+  FormItem,
+  FormLabel,
+  FormMessage,
+  FormRootError,
+  useFieldArray,
+  useForm,
+} from "@ctrlplane/ui/form";
+import { Input } from "@ctrlplane/ui/input";
+import { Label } from "@ctrlplane/ui/label";
+
+import { api } from "~/trpc/react";
+import { TargetConfigEditor } from "./TargetConfigEditor";
+
+type TargetWithMetadata = Target & {
+  metadata: Record<string, string>;
+};
+
+const editTargetSchema = z.object({
+  name: z.string(),
+  kind: z.string(),
+  identifier: z.string().min(4),
+  version: z.string(),
+  config: z.string().refine(
+    (val) => {
+      try {
+        const output = yaml.load(val);
+        const isValidRecord = z.record(z.any()).safeParse(output).success;
+        return isValidRecord;
+      } catch {
+        return false;
+      }
+    },
+    { message: "Config must be a valid YAML Object" },
+  ),
+  metadata: z
+    .array(
+      z.object({
+        key: z.string().min(1, "Key is required"),
+        value: z.string().min(1, "Value is required"),
+      }),
+    )
+    .refine(
+      (arr) => {
+        const keys = arr.map((item) => item.key);
+        return keys.length === new Set(keys).size;
+      },
+      { message: "Metadata keys must be unique" },
+    ),
+});
```

Comment:
Felt like I should of been able to just reference
```tsx
z.record().string()
```
and it should of worked, but I was having problems with the 

```tsx
 const { fields, append, remove } = useFieldArray({
    name: "metadata",
    control: form.control,
  });
  ```